lib/conversion.js

```diff
@@ -53,7 +53,7 @@ function convert(conversionOptions, cb) {
         throw new Error('can\'t use waitForJS option if settings.javascriptEnabled is not activated');
     }
 
-    var id = uuid();
+    var id = opt.filename || uuid();
```

Comment:
i would name this option as `tmpId`, since `filename` usually implies that the value has a file extension, which in this case we don't expect a value in that format.

Reply:
I am ok with that, the name does not really matter to me here, it's the functionality that i am concerned about.

Reply:
I mean, ideally we should be specifying the filename including extension, right now '.pdf' is hardcoded in the package itself which i think can be removed if we are using 'filename' option. In some cases temporary files are not required to have an extension since we can move them later to their final destination.

I am usually against hard coded values (extensions, names etc..) the ability to customize/override those options are a necessity for some.

Thank You.

Reply:
> I mean, ideally we should be specifying the filename including extension, right now '.pdf' is hardcoded in the package itself which i think can be removed if we are using 'filename' option. In some cases temporary files are not required to have an extension since we can move them later to their final destination.

in this case we use the `id` not just for the final pdf file, but also for writing to temporal html file, having the `tmpId` instead of a filename is more general and practical for this case. anyway all of this is an implementation detail of the library, i'm not sure why it is important to customize a filename of a temporal file that it is just used internally. you have total control of the name of your final pdf file, since you receive a stream and content of the pdf in node.js callback, the temp file (that is an implementation detail) is not important.

> I am usually against hard coded values (extensions, names etc..) the ability to customize/override those options are a necessity for some.

i agree with that, but in this case we are talking about a file that is an implementation detail of the library, and also there are other files written when the library does its job, should we have an option to allow to override the name of every file in the internal implementation? i don't think that is a good idea

Reply:
In this case the temp file was important since it was overwriting each other due to the race condition by using the timestamp, if we had the ability to set the temporary file we could've easily set it to a more unique name so it won't overwrite, that's how this whole thing started.

Having the ability to customize that file just gives us more control, we don't do anything with the file, we use the temporary file since it's already written and ready to go it saves time and resources moving it to a different location.

Other files that are being written temporarily do not matter (or don't matter as much) since the temporary pdf file is the one that we need and use.

This can be argued in lots of ways, i've already forked this repository and made the necessary changes to be able to have little bit more control over the temporary file that is generated, i PRed it just in case since most likely this will come up for other users as it's an issue with the package itself (regarding the race condition) but either way i am ok with using the fork if this won't get merged, it's not a big deal. 😀

Thanks for help. 👍🏻

Reply:
> In this case the temp file was important since it was overwriting each other due to the race condition by using the timestamp

yes, but switching to v4 (as this PR does) solves the problem. having an option to control the internal id of the temp file is just a nice workaround when something like this problem happens.

> Having the ability to customize that file just gives us more control, we don't do anything with the file, we use the temporary file since it's already written and ready to go it saves time and resources moving it to a different location.

i understand that but it is not the way in which this library was mean to be used, you are doing just something special, i'm not even sure how you are getting the temp file associated with a particular rendering request, seems like you are just watching the temp folder and moving every file that you find in it, which has other types of downsides.

> Other files that are being written temporarily do not matter (or don't matter as much) since the temporary pdf file is the one that we need and use.

yes, but do you realize that you are just solving the problem for your case? 😃 it would be weird to have an option that just customize one temporal file, we use an id because it lets us create a filename depending of the type of the file that we need internally, having a `filename` for just one temp file would be weird because it would mean that other files would have different names. we value the fact the temp files in a rendering request are named like [id].[type of content]. that allows us to quickly know how many files are associated in a particular case.  




> This can be argued in lots of ways, i've already forked this repository and made the necessary changes to be able to have little bit more control over the temporary file that is generated, i PRed it just in case since most likely this will come up for other users as it's an issue with the package itself (regarding the race condition) but either way i am ok with using the fork if this won't get merged, it's not a big deal. 😀

thanks for creating the issue, i'm just trying to give you my perspective as a maintainer of the library 😃. 

i would say that we are going to accept the PR if you make the change that i requested, i think i gave you all the explanation for that change.. if you are still interested in the PR please do it, that is how open source works 😄

Reply:
i just made that change. thanks again.